### PR TITLE
migration-from-draft-js

### PR DESCRIPTION
**Pull request by akanshaaa19**
## Summary

- [x] Chat Page Text editor
- [x] template text editor
- [x] interactive mesages text editor
- [x] emoji input
- [x] mentions plugin
- [x] styling
- [x] clean up

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     "@emotion/react": "^11.11.3",
     "@emotion/styled": "^11.11.0",
     "@glific/flow-editor": "^1.19.3-5",
+    "@lexical/react": "^0.13.1",
     "@mui/icons-material": "^5.15.6",
```
Comment (mdshamoon) [inline package.json:24]: Remove draft js and plugins from package.json

File: src/common/RichEditor.tsx [hunk 2]
```diff
@@ -9,9 +9,9 @@ const regexForLink =
   /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_+.~#?&/=]*)/gi;
 
 // Convert Draft.js to WhatsApp message format.
-export const getPlainTextFromEditor = (editorState: any) =>
-  editorState.getCurrentContent().getPlainText();
-
+export const getPlainTextFromEditor = (editorState: any) => {
+  return editorState;
+};
```
Comment (mdshamoon) [inline src/common/RichEditor.tsx:14]: If its directly returning the editor state I don't think we need this function.

File: src/components/UI/EmojiPicker/EmojiPicker.tsx [hunk 3]
```diff
@@ -1,15 +1,30 @@
...
   displayStyle: object;
   onEmojiSelect: Function;
 }
 
-export const EmojiPicker = ({ displayStyle, onEmojiSelect }: EmojiPickerProps) => (
-  <div data-testid="emoji-container" style={displayStyle}>
-    <Picker data={data} onEmojiSelect={onEmojiSelect} />
-  </div>
-);
+export const EmojiPicker = ({ displayStyle, onEmojiSelect }: EmojiPickerProps) => {
```
Comment (mdshamoon) [inline src/components/UI/EmojiPicker/EmojiPicker.tsx:10]:
```suggestion
export const EmojiPicker = ({ displayStyle }: EmojiPickerProps) => {
```
Remove this if not needed. 
Also I am not sure if we should tightly couple editor with the picker. There could be a case where the emoji picker is needed without the editor. Let's discuss this once

File: src/components/UI/Form/EmojiInput/EmojiInput.tsx [hunk 4]
```diff
@@ -83,111 +28,31 @@ export const EmojiInput = ({
...
-      if (newState) {
-        props.form.setFieldValue(name, newState);
-        return 'handled';
-      }
-    }
-    return 'not-handled';
-  };
-
-  const draftJsChange = (editorState: any) => {
+  const lexicalChange = (editorState: any) => {
```
Comment (mdshamoon) [inline src/components/UI/Form/EmojiInput/EmojiInput.tsx:31]:
```suggestion
  const lexicalChange = (editorState: string) => {
```
Let's add types where we know

File: src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.module.css [hunk 5]
```diff
@@ -28,3 +28,29 @@ span.emoji-mart-emoji {
...
+  resize: none;
+  font-size: 16px;
+  overflow: auto;
+  width: 100%;
+  min-height: 40px;
+  margin-top: 4px;
+  position: relative;
+  tab-size: 1;
+  outline: 0;
+  /* padding: 4px 10px; */
```
Comment (mdshamoon) [inline src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.module.css:42]: Remove if not needed

File: src/containers/Chat/ChatMessages/ChatInput/ChatInput.tsx [hunk 6]
```diff
@@ -484,18 +500,9 @@ export const ChatInput = ({
...
-                }
+                submitMessage(editorState);
               }}
-              disabled={
-                (!editorState.getCurrentContent().hasText() &&
-                  !attachmentAdded &&
-                  !recordedAudio) ||
-                uploading
-              }
+              // disabled={(!attachmentAdded && !recordedAudio) || uploading}
```
Comment (mdshamoon) [inline src/containers/Chat/ChatMessages/ChatInput/ChatInput.tsx:505]: Please check if this needs to commented out.

File: src/containers/InteractiveMessage/InteractiveMessage.helper.ts [hunk 7]
```diff
@@ -39,7 +39,7 @@ export const validator = (templateType: any, t: any) => {
       .required(t('Title is required'))
       .max(60, t('Title can be at most 60 characters')),
     body: Yup.string()
-      .transform((_current, original) => original.getCurrentContent().getPlainText())
+      // .transform((_current, original) => original.getCurrentContent().getPlainText())
```
Comment (mdshamoon) [inline src/containers/InteractiveMessage/InteractiveMessage.helper.ts:42]: Remove of not needed. Please check other places as well

**Review by mdshamoon — CHANGES_REQUESTED** [review]

**Comment by mdshamoon** [comment]
Testing feedbacks:

1. In templates page, the text moves out of the box if I type a long message
image

2. Mentions plugin seems a bit buggy in interactive messages. Not able to select using keyboard arrows and enter button. Also it moves along with scroll
image

3. Pressing shift + enter should move to a new line and not send a message on chat screen
image

**Comment by akanshaaa19** [comment]
@mdshamoon resolved these

File: src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.tsx [hunk 8]
```diff
@@ -1,110 +1,115 @@
...
-  const handleKeyCommand = (command: string, editorStateChange: any) => {
-    // On enter, submit. Otherwise, deal with commands like normal.
-    if (command === 'enter') {
-      // Convert Draft.js to WhatsApp
-      sendMessage(getPlainTextFromEditor(editorStateChange));
-      return 'handled';
-    }
+  const onResize = useCallback((height: any) => {
+    handleHeightChange(height - 40);
+  }, []);
```
Comment (mdshamoon) [inline src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.tsx:38]: Check if we even need this. If not lets remove it.

File: src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.tsx [hunk 9]
```diff
@@ -1,110 +1,115 @@
...
+  const handleFormatting = (text: string, formatter: string) => {
+    switch (formatter) {
+      case 'bold':
+        return `*${text}*`;
+      case 'italic':
+        return `_${text}_`;
+      case 'strikethrough':
+        return `~${text}~`;
+      default:
+        return text;
```
Comment (mdshamoon) [inline src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.tsx:62]: This is used in multiple places. Move this to utils and import everywhere.

File: src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.tsx [hunk 10]
```diff
@@ -1,110 +1,115 @@
...
+        if (event.code === 'Enter' && !event.shiftKey) {
+          event.preventDefault();
+          let textMessage = getTextContent(editor);
+          sendMessage(textMessage);
+          return true;
+        } else if ((event.ctrlKey || event.metaKey) && event.code === 'KeyB') {
+          formatter = 'bold';
+        } else if ((event.ctrlKey || event.metaKey) && event.code === 'KeyI') {
+          formatter = 'italic';
+        }
```
Comment (mdshamoon) [inline src/components/UI/Form/WhatsAppEditor/WhatsAppEditor.tsx:81]: This part is also common and can be reused. Move this to utils and import here

File: src/components/UI/MessageDialog/MessageDialog.tsx [hunk 11]
```diff
@@ -21,13 +22,20 @@ export const MessageDialog = ({ title, onSendMessage, handleClose }: MessageDial
...
-      />
+      <LexicalComposer
+        initialConfig={{
+          namespace: 'chat-input',
+          onError: (error: any) => console.log(error),
+        }}
+      >
+        <ChatInput
+          onSendMessage={onSendMessage}
+          handleHeightChange={() => {}}
```
Comment (mdshamoon) [inline src/components/UI/MessageDialog/MessageDialog.tsx:33]: Check if this is needed?

File: src/containers/Chat/ChatMessages/ChatInput/ChatInput.test.tsx [hunk 12]
```diff
@@ -170,7 +179,14 @@ describe('<ChatInput />', () => {
...
-        <ChatInput {...propsWithBspStatusNone} />
+        <LexicalComposer
+          initialConfig={{
+            namespace: 'chat-input',
+            onError: (error: any) => console.log(error),
+          }}
+        >
+          <ChatInput {...propsWithBspStatusNone} />
+        </LexicalComposer>
       </MockedProvider>
```
Comment (mdshamoon) [inline src/containers/Chat/ChatMessages/ChatInput/ChatInput.test.tsx:190]: This seems to be common in all the test cases. Create a function for this and reuse.
```suggestion
   const chatInput= (props) =>
        <LexicalComposer
          initialConfig={{
            namespace: 'chat-input',
            onError: (error: any) => console.log(error),
          }}
        >
          <ChatInput {...props} />
        </LexicalComposer>
      </MockedProvider>
```

File: src/containers/Chat/ChatMessages/ChatMessages.tsx [hunk 13]
```diff
@@ -595,7 +593,7 @@ export const ChatMessages = ({ contactId, collectionId, startingHeight }: ChatMe
     messageListContainer = (
       <Container
         className={`${styles.MessageList} messageContainer `}
-        style={{ height: `calc(100% - 195px - ${reducedHeight}px)` }}
+        style={{ height: `calc(100% - 215px ` }}
```
Comment (mdshamoon) [inline src/containers/Chat/ChatMessages/ChatMessages.tsx:596]: Any reason we increased the height?

File: src/containers/Template/Form/SpeedSend/SpeedSend.test.tsx [hunk 14]
```diff
@@ -6,13 +7,34 @@ import { SpeedSendList } from 'containers/Template/List/SpeedSendList/SpeedSendL
...
+  writable: true,
+  value: (query: any) => {
+    return {
+      matches: false,
+      media: query,
+      onchange: null,
+      addListener: vi.fn(),
+      removeListener: vi.fn(),
+    };
+  },
```
Comment (mdshamoon) [inline src/containers/Template/Form/SpeedSend/SpeedSend.test.tsx:35]: Do we need this. I think we already imported it.

**Review by mdshamoon — CHANGES_REQUESTED** [review]
Overall functionality wise it looks good. Some minor tweeks needed.

**Review by mdshamoon — APPROVED** [review]

**Pull request merged**